File: P1/Ejercicio1/carrera/Carrera.java
package carrera;

import bicicleta.Bicicleta;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.Random;

/**
 * Clase abstracta para representar la carrera de bicicletas. Esta implementa la interfaz Runnable para poder ejecutar
 * carreras de manera concurrente
 */
public abstract class Carrera implements Runnable {
    /**
     * Contendrá todas las bicicletas que participarán en la carrera
     */
    protected ArrayList<Bicicleta> bicicletas;

    /**
     * Número de bicicletas retiradas de la carrera
     */
    protected int bicicletasRetiradas;


    // Métodos NO abstractos:

    /**
     * Constructor por parámetro que inicializa tanto el número de bicicletas retiradas como el array de las bicicletas
     * para poder utilizarlo sin problemas a partir del momento en el que se cree la instancia
     * @param numBicicletas Número de bicicletas que participarán en la carrera
     */
    public Carrera(int numBicicletas) {
        bicicletas          = new ArrayList<>(numBicicletas);
        bicicletasRetiradas = 0;
    }

    /**
     * Añade la bicicleta pasada al array de bicicletas de la carrera
     * @param bicicleta Bicicleta a añadir a la carrera
     */
    public void aniadirBicicleta(Bicicleta bicicleta) {
        bicicletas.add(bicicleta);
    }

    /**
     * Retira una bicicleta escogida al azar de la carrera
     */
    public void retirarBicicletaAleatoria() {
        // Verificar si la lista no está vacía:
        if (!bicicletas.isEmpty()) {
            // Crear un objeto Random:
            Random rand = new Random();

            // Generar un índice aleatorio:
            int indiceAleatorio = rand.nextInt(bicicletas.size());
            int idBicicleta     = bicicletas.get(indiceAleatorio).obtenerId();

            // Eliminar el elemento en el índice aleatorio:
            bicicletas.remove(indiceAleatorio);

            // Informar de la retirada de la bicicleta en cuestión:
            System.out.println("Retirada bicicleta '" + idBicicleta + "'");
        } else {
            // En este punto, no hay bicicletas en la carrera:
            System.out.println("No hay bicicletas en la carrera");
        }
    }

    /**
     * Retira un número de bicicletas pasado de la carrera
     * @param numBicicletas Número de bicicletas a retirar de la carrera
     */
    private void retirarBicicletas(int numBicicletas) {
        for (int i = 0; i < numBicicletas; i++) {
            retirarBicicletaAleatoria();
        }
    }

    /**
     * Calcula y almacena el número de bicicletas que serán retiradas de la carrera antes de que esta finalice
     * @param porcentaje Porcentaje de bicicletas que se retirarán de la carrera
     */
    public void computarRetiradas(int porcentaje) {
        bicicletasRetiradas = (bicicletas.size() * porcentaje) / 100;
    }

    /**
     * Consultor del identificador de la bicicleta que ha ganado la carrera
     * @return Identificador de la bicicleta que ha ganado la carrera
     */
    public int obtenerIdGanador() {
        this.bicicletas.sort(Comparator.comparingInt(o -> o.distanciaTotal));

        return (this.bicicletas.getLast().obtenerId());
    }

    /**
     * Simula la carrera. Para ello, la inicia y hace que las bicicletas avancen mientras dura la carrera, haciendo que
     * se retiren algunas, finalizando la carrera al final
     */
    public void simular() {
        final long tiempoCarrera = 60000; // 60 segundos

        // Iniciar carrera:
        iniciarCarrera();
        long inicioTiempo = System.currentTimeMillis();

        // Mientras no se agote el tiempo de carrera estipulado, las bicicletas permanecerán avanzando:
        while (System.currentTimeMillis() - inicioTiempo < tiempoCarrera) {
            // Avanza cada bicicleta n metros aleatorios:
            for (Bicicleta bici : bicicletas) {
                bici.avanzar();
            }

            try {
                // Pausa la ejecución del hilo durante 1 segundo (esto facilita el seguimiento de la simulación):
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

        // Retirar bicicletas aleatorias antes de finalizar la carrera:
        retirarBicicletas(bicicletasRetiradas);

        // Finalizar carrera y retirada del resto de bicicletas:
        finalizarCarrera();
        retirarBicicletas(bicicletas.size());
    }


    // Métodos abstractos:

    /**
     * Inicia la carrera de bicicletas
     */
    public abstract void iniciarCarrera();

    /**
     * Finaliza la carrera de bicicletas
     */
    public abstract void finalizarCarrera();
}

File: P1/Ejercicio2/Carreras/carrera.py
from abc import ABC, abstractmethod
from secrets import choice
from typing import List
from threading import *
from Bicicletas.bicicleta import *

class Carrera(ABC):
    """
    Clase abstracta Carrera que define la estructura básica de una carrera.
    Atributos:
        _bicicletas (List[Bicicleta]): Lista de bicicletas participantes en la carrera.
        _hilos (List[Thread]): Lista de hilos para simular el avance de cada bicicleta.
        _continuar (Event): Evento que permite a las bicicletas avanzar.
    """
    
    def __init__(self) -> None:
        """
        Constructor de la clase Carrera que inicializa las listas de bicicletas y hilos.
        """
        self._bicicletas : List[Bicicleta] = []
        self._hilos = []
        self._continuar = Event()
        self._continuar.set()  # Inicialmente permitimos que las bicicletas avancen
        
    def aniadir_bicicleta(self, bicicleta: Bicicleta) -> None:
        """
        Añade una bicicleta a la carrera y crea un hilo para su avance.
        
        Parámetros:
            bicicleta (Bicicleta): La bicicleta a añadir a la carrera.
        """
        self._bicicletas.append(bicicleta)
        hilo = Thread(target=bicicleta.avanzar, args=(self._continuar,))
        self._hilos.append(hilo)
        
    def detener_bicicletas(self, tipo) -> None:
        """
        Detiene las bicicletas.
        Retira las bicicletas.
        """
        self._continuar.clear()  # Indica a las bicicletas que deben detenerse
        if tipo == 'carretera':
            self.retirar_bicicletas(0.10, tipo) # Una vez detenidas, proceder a retirar bicicletas
        else:
            self.retirar_bicicletas(0.20, tipo)
    
    
    def retirar_bicicletas(self, porcentaje, tipo_carrera):
        """
        Retira un porcentaje de bicicletas de la carrera.
        
        Parámetros:
            porcentaje (float): Porcentaje de bicicletas a retirar.
            tipo_carrera (str): Tipo de carrera para mostrar en el mensaje.
        """
        numero_a_retirar = int(len(self._bicicletas) * porcentaje)
        print(f"Bicicletas retiradas de la carrera de {tipo_carrera}:\n")
        for _ in range(numero_a_retirar):
          self.retirar_bicicleta_aleatoria(tipo_carrera)
          
    def retirar_bicicleta_aleatoria(self, tipo_carrera) -> None:
        """
        Retira una bicicleta aleatoria de la carrera.
        """
        # Verifica si la lista no esta vacia
        if len(self._bicicletas) > 0:
            # Escoge una bicicleta aleatoria y la elimina
            bicicleta : Bicicleta = choice(self._bicicletas)
            self._bicicletas.remove(bicicleta)
            print(f"Se ha retirado la bicicleta con identificador \'{bicicleta._id}\' de la carrera de {tipo_carrera}")
        else:
            print("No hay bicicletas en la carretera")
    
    @abstractmethod
    def iniciar_carrera(self) -> None: 
        """Método abstracto para iniciar la carrera."""
        pass
    
    @abstractmethod
    def finalizar_carrera(self) -> None: 
        """Método abstracto para finalizar la carrera."""
        pass
    
    @abstractmethod
    def ganador_carrera(self) -> None: 
        """Método abstracto para determinar el ganador de la carrera."""
        pass